activerecord/test/models/clothing_item.rb

```diff
@@ -2,6 +2,8 @@
 
 class ClothingItem < ActiveRecord::Base
   query_constraints :clothing_type, :color
+
+  has_many :clothing_item_inventories, foreign_key: [:clothing_type, :clothing_item_color], primary_key: [:clothing_type, :color]
```

Comment:
I've been looking at this and I don't yet have a solution but it _feels like_ the `primary_key` and `foreign_key` declarations are basically the same which makes for a pretty awkward feeling API.

I hope that we can find an API that we only need to define once on an association and then infer the remaining information from that. I was thinking originally something like this:

```
has_many : clothing_item_inventories, query_constraints: [:clothing_type, :color]
```

But I know associations are really complex so this might not be nearly as possible as I was hoping.

Reply:
>I hope that we can find an API that we only need to define once on an association and then infer the remaining information from that.

That's what I'm hoping to reach as well. For example instead of deriving `primary_key(active_record)` here
https://github.com/rails/rails/blob/1dc878b84b702d3ff837f09ed851ab7a60ef01c5/activerecord/lib/active_record/reflection.rb#L490
we would derive the value from the query constraints list
and maybe even make `derive_foreign_key` smart enough to allow the API as simple as `has_many :clothing_item_inventories` 
https://github.com/rails/rails/blob/1dc878b84b702d3ff837f09ed851ab7a60ef01c5/activerecord/lib/active_record/reflection.rb#L473

I'm going to switch to a different models to be used in tests so we can easier see what is possible